### PR TITLE
OKAPI-1053: Add module users section to raml schema

### PR DESCRIPTION
**Pull request by MikhailFokanov**
As discussed (Jakub Skoczen, Vince Bareau, Craig McNally) module descriptor should have section for module user metadata. This section should contain user type and the permissions, that should be assigned for the user.

Acceptance criteria:

The section "user" is added to the raml schema
The section contains "type" and "permissions" property

### CODE REVIEW

File: okapi-core/src/main/raml/ModuleDescriptor.json [hunk 1]
```diff
@@ -80,6 +80,10 @@
     "launchDescriptor": {
       "description": "Default deployment for this module",
       "$ref": "LaunchDescriptor.json"
+    },
+    "user" : {
+      "description": "Module user descriptor",
+      "$ref": "LaunchDescriptor.json"
```
Comment (hjiebsco) [inline okapi-core/src/main/raml/ModuleDescriptor.json:86]: Wrong `$ref`

File: okapi-core/src/main/raml/ModuleUser.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "LaunchDescriptor.json",
+  "title": "LaunchDescriptor",
+  "description": "Specifies how a module is deployed and undeployed",
```
Comment (hjiebsco) [inline okapi-core/src/main/raml/ModuleUser.json:5]: Copy/paste error. Does not match file name `ModuleUser.json`.

File: okapi-core/src/main/raml/ModuleDescriptor.json (hunk 1 above)
Comment (MikhailFokanov-ebsco) [reply okapi-core/src/main/raml/ModuleDescriptor.json]: oops. Fixed.

File: okapi-core/src/main/raml/ModuleUser.json (hunk 2 above)
Comment (MikhailFokanov-ebsco) [reply okapi-core/src/main/raml/ModuleUser.json]: fixed

Comment (MikhailFokanov-ebsco) [reply okapi-core/src/main/raml/ModuleUser.json]: fixed

File: okapi-core/src/main/raml/ModuleUser.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "LaunchDescriptor.json",
+  "title": "LaunchDescriptor",
+  "description": "Specifies the user that should be created for the module",
+  "type": [ "object", "null" ],
+  "additionalProperties" : false,
+  "properties": {
+    "type": {
+      "description": "The type of the module user",
+      "type": ["string"]
```
Comment (craigmcnally) [inline okapi-core/src/main/raml/ModuleUser.json:11]: Should this be an enum?

**Review by adamdickmeiss — CHANGES_REQUESTED** [review]
I have no idea about the semantics of the new structure. When happy, be aware that you must create POJOs for it yourself in Okapi as code is NOT generated from schemas.  See ModuleDescriptpr.json. You'd probably also want a unit test where you post a module with this info to verify.

**Comment by adamdickmeiss** [comment]
> As discussed (Jakub Skoczen, Vince Bareau, Craig McNally) module descriptor should have section for module user metadata. This section should contain user type and the permissions, that should be assigned for the user.
> 
> Acceptance criteria:
> 
> The section "user" is added to the raml schema The section contains "type" and "permissions" property

If the intention of this is *metadata* and NOT managed by Okapi , then why don't you use the "metadata" property of Okapi to hold this. This is "any" object and can hold whatever you like.

**Comment by MikhailFokanov** [comment]
> > As discussed (Jakub Skoczen, Vince Bareau, Craig McNally) module descriptor should have section for module user metadata. This section should contain user type and the permissions, that should be assigned for the user.
> > Acceptance criteria:
> > The section "user" is added to the raml schema The section contains "type" and "permissions" property
> 
> If the intention of this is _metadata_ and NOT managed by Okapi , then why don't you use the "metadata" property of ModuleDescriptor to hold this. This is "any" object and can hold whatever you like.

@craigmcnally do you like the idea of adding "user" to "metadata" property?

**Comment by craigmcnally** [comment]
> > > As discussed (Jakub Skoczen, Vince Bareau, Craig McNally) module descriptor should have section for module user metadata. This section should contain user type and the permissions, that should be assigned for the user.
> > > Acceptance criteria:
> > > The section "user" is added to the raml schema The section contains "type" and "permissions" property
> > 
> > 
> > If the intention of this is _metadata_ and NOT managed by Okapi , then why don't you use the "metadata" property of ModuleDescriptor to hold this. This is "any" object and can hold whatever you like.
> 
> @craigmcnally do you like the idea of adding "user" to "metadata" property?

Would this mean that it couldn't be structured / controlled by a schema?  I see metadata is just "object" w/o a schema.

**Comment by craigmcnally** [comment]
@adamdickmeiss is there any precedent for how metadata is used in the module descriptor?  Is it currently being used for anything?

**Comment by adamdickmeiss** [comment]
> @adamdickmeiss is there any precedent for how metadata is used in the module descriptor? Is it currently being used for anything?

AFAIK no. Of course, the drawback is that the structure is undefined, so if you wish some structure it's no good.

**Comment by craigmcnally** [comment]
IMO having a well defined structure for this is desirable, possibly including an enumeration of valid values (I can probably be dissuaded of that part though)

File: okapi-core/src/main/raml/ModuleUser.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    },
+    "permissions": {
+      "description": "Permission ID (usually module.service.method or similar)",
+      "type": ["array", "null"],
+      "items": {
+        "type": "string",
+        "enum": ["module", "tenant"],
+        "default": "module"
+      }
```
Comment (adamdickmeiss) [inline okapi-core/src/main/raml/ModuleUser.json:20]: Why is this an enum and not the "type" earlier.. I thought the purpose of this array was to list the permissions needed for that user. If we don't, how do we know what the module needs?

File: okapi-core/src/main/raml/ModuleUser.json (hunk 3 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/raml/ModuleUser.json:11]: Yes. But not yet.

**Review by adamdickmeiss — CHANGES_REQUESTED** [review]

File: okapi-core/src/main/raml/ModuleDescriptor.json [hunk 5]
```diff
@@ -80,6 +80,10 @@
     "launchDescriptor": {
       "description": "Default deployment for this module",
       "$ref": "LaunchDescriptor.json"
+    },
+    "user" : {
```
Comment (craigmcnally) [inline okapi-core/src/main/raml/ModuleDescriptor.json:84]: Do we foresee the need to ever define more than one module user?

Comment (adamdickmeiss) [reply okapi-core/src/main/raml/ModuleDescriptor.json:84]: How would those be distinguished from one another? This is per module it seems.

File: okapi-core/src/main/raml/ModuleUser.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  "additionalProperties" : false,
+  "properties": {
+    "type": {
+      "description": "The type of the module user",
+      "type": "string",
+      "enum": ["module", "tenant"],
+      "default": "module"
+    },
+    "permissions": {
+      "description": "Permission ID (usually module.service.method or similar)",
```
Comment (adamdickmeiss) [inline okapi-core/src/main/raml/ModuleUser.json:16]: I don't understand this description. Isn't this the permissions that the module user requires to operate?

**Review by adamdickmeiss — APPROVED** [review]

File: okapi-core/src/main/java/org/folio/okapi/bean/ModuleUser.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package org.folio.okapi.bean;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Tells how a module is to be deployed. Either by exec'ing a command (and
+ * killing the process afterwards), or by using command lines to start and stop
+ * it.
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/bean/ModuleUser.java:10]: Copy & paste from https://github.com/folio-org/okapi/blob/v4.12.0/okapi-core/src/main/java/org/folio/okapi/bean/LaunchDescriptor.java , needs changes for ModuleUser.

File: okapi-core/src/main/raml/ModuleDescriptor.json (hunk 5 above)
Comment (julianladisch) [reply okapi-core/src/main/raml/ModuleDescriptor.json:84]: I also suggest multiple module users with different permissions. For example a migration might require more permissions than the regular work.
We can easily add some identifier to distinguish them.

Comment (MikhailFokanov) [reply okapi-core/src/main/raml/ModuleDescriptor.json:84]: The additional deployment complexity of managing identifiers for users and providing of multiples usernames along with credentials to the module environment variables without corresponding business requirements seems to be a premature optimization. It is also inconvenient to provide Map<String,String> of arbitrary size (username and pass for each user) in environment variable.

File: okapi-core/src/main/java/org/folio/okapi/bean/ModuleUser.java (hunk 7 above)
Comment (MikhailFokanov) [reply okapi-core/src/main/java/org/folio/okapi/bean/ModuleUser.java:10]: done

File: okapi-core/src/main/raml/ModuleUser.json (hunk 6 above)
Comment (MikhailFokanov) [reply okapi-core/src/main/raml/ModuleUser.json:16]: done

File: okapi-core/src/main/raml/ModuleUser.json (hunk 4 above)
Comment (MikhailFokanov) [reply okapi-core/src/main/raml/ModuleUser.json:20]: done

**Review by adamdickmeiss — CHANGES_REQUESTED** [review]
Close this without merge?

**Pull request closed without merging**